pkg/server/plugin/keymanager/gcpkms/fetcher.go

```diff
@@ -0,0 +1,166 @@
+package gcpkms
+
+import (
+	"context"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	keymanagerv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/keymanager/v1"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/iterator"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type keyFetcher struct {
+	keyRing        string
+	kmsClient      kmsClient
+	listCryptoKeys listCryptoKeysFn
+	log            hclog.Logger
+	serverID       string
+	tdHash         string
+}
+
+// fetchKeyEntries requests Cloud KMS to get the list of CryptoKeys that are
+// active in this server. They are returned as a keyEntry array.
+func (kf *keyFetcher) fetchKeyEntries(ctx context.Context) ([]*keyEntry, error) {
+	var keyEntries []*keyEntry
+	var keyEntriesMutex sync.Mutex
+	g, ctx := errgroup.WithContext(ctx)
+
+	it := kf.listCryptoKeys(ctx, kf.kmsClient, &kmspb.ListCryptoKeysRequest{
+		Parent: kf.keyRing,
+		Filter: fmt.Sprintf("labels.%s = %s AND labels.%s = %s AND labels.%s = true",
+			labelNameServerTD, kf.tdHash, labelNameServerID, kf.serverID, labelNameActive),
+	})
+	for {
+		cryptoKey, err := it.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to list SPIRE Server keys in Cloud KMS: %v", err)
+		}
+		spireKeyID, ok := kf.getSPIREKeyIDFromCryptoKeyName(cryptoKey.Name)
```

Comment:
`buildKeyEntryFromCryptoKey` receives `cryptoKey` as an argument, so it checks that it's not nil just in case the caller call the function with an invalid value.